lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+  });
+
+  // Update the operations in the module.
+  SmallVector<Operation *> deleteOps;
+  SmallVector<std::pair<Value, Value>> connects;
+  module.walk([&](Operation *op) {
+    implementAsyncReset(op, module, actualReset, deleteOps, connects);
+  });
+
+  // Remove the obsolete instances.
+  for (auto op : deleteOps)
+    op->erase();
+
+  // Add the necessary connects.
+  OpBuilder builder(module);
+  builder.setInsertionPointToEnd(module.getBodyBlock());
+  for (auto con : connects)
+    builder.create<ConnectOp>(con.first.getLoc(), con.first, con.second);
+
+  return success();
+}
+
+/// Modify an operation in a module to implement an async reset for that module.
+void InferResetsPass::implementAsyncReset(
+    Operation *op, FModuleOp module, Value actualReset,
+    SmallVectorImpl<Operation *> &deleteOps,
+    SmallVectorImpl<std::pair<Value, Value>> &connects) {
+  ImplicitLocOpBuilder builder(op->getLoc(), op);
+
+  // Handle instances.
+  if (auto instOp = dyn_cast<InstanceOp>(op)) {
+    // Lookup the reset domain of the instantiated module. If there is no reset
+    // domain associated with that module, or the module is explicitly marked as
+    // being in no domain, simply skip.
+    auto domainIt = domains.find(instOp.getReferencedModule());
+    if (domainIt == domains.end())
+      return;
+    auto &domain = domainIt->second.back().first;
+    if (!domain.reset)
+      return;
+    LLVM_DEBUG(llvm::dbgs() << "- Update instance '" << instOp.name() << "'\n");
+
+    // If needed, add a reset port to the instance.
+    Value instReset;
+    if (domain.newPortName) {
+      LLVM_DEBUG(llvm::dbgs() << "  - Adding new result as reset\n");
+
+      // Determine the new result types.
+      SmallVector<Type> resultTypes;
+      resultTypes.reserve(instOp.getNumResults() + 1);
+      resultTypes.push_back(actualReset.getType());
+      resultTypes.append(instOp.getResultTypes().begin(),
+                         instOp.getResultTypes().end());
+
+      // Create a new list of port annotations.
+      ArrayAttr newPortAnnos;
+      if (auto oldPortAnnos = instOp.portAnnotations()) {
+        SmallVector<Attribute> buffer;
+        buffer.reserve(oldPortAnnos.size() + 1);
+        buffer.push_back(builder.getArrayAttr({}));
+        buffer.append(oldPortAnnos.begin(), oldPortAnnos.end());
+        newPortAnnos = builder.getArrayAttr(buffer);
+      } else {
+        newPortAnnos = builder.getArrayAttr({});
+      }
+
+      // Create a new instance op with the reset inserted.
+      auto newInstOp = builder.create<InstanceOp>(
+          resultTypes, instOp.moduleName(), instOp.name(), instOp.annotations(),
+          newPortAnnos);
+      instReset = newInstOp.getResult(0);
+
+      // Update the uses over to the new instance and drop the old instance.
+      for (unsigned i = 0, e = instOp.getNumResults(); i < e; ++i)
+        instOp.getResult(i).replaceAllUsesWith(newInstOp.getResult(i + 1));
```

Comment:
Operation has a replaceAllUsesWith method that takes a range that might simplify this, see ~ mlir/IR/Operation.h:152

Reply:
Neat! Changed that.